File: Packages/TeamTasksKit/Sources/TeamTasksMocks/MockClock.swift
import Foundation
import TeamTasksCore

/// Controllable clock for tests, previews and UI tests.
///
/// `autoAdvance` adds a fixed amount after every read so that successive backend operations get strictly
/// increasing timestamps (like successive Postgres transactions), while staying fully deterministic.
public final class MockClock: Sendable {
    private let state: LockedValue<Date>
    private let step: TimeInterval

    public init(_ start: Date, autoAdvance: TimeInterval = 0) {
        state = LockedValue(start)
        step = autoAdvance
    }

    /// Returns the current date, then advances by `autoAdvance`.
    public func now() -> Date {
        state.withValue { value in
            let current = value
            value = value.addingTimeInterval(step)
            return current
        }
    }

    /// The current date, without advancing.
    public func peek() -> Date {
        state.withValue { $0 }
    }

    public func advance(by seconds: TimeInterval) {
        state.withValue { $0 = $0.addingTimeInterval(seconds) }
    }

    public func set(_ date: Date) {
        state.withValue { $0 = date }
    }

    /// A `NowProvider` reading this clock.
    public var provider: NowProvider {
        { [self] in now() }
    }
}

/// Minimal lock-protected box (Synchronization.Mutex needs iOS 18).
final class LockedValue<Value>: @unchecked Sendable {
    private let lock = NSLock()
    private var value: Value

    init(_ value: Value) {
        self.value = value
    }

    func withValue<Result>(_ body: (inout Value) throws -> Result) rethrows -> Result {
        lock.lock()
        defer { lock.unlock() }
        return try body(&value)
    }
}

File: android/mocks/src/main/kotlin/io/github/notkanaa/equipe/mocks/MockClock.kt
package io.github.notkanaa.equipe.mocks

import io.github.notkanaa.equipe.core.NowProvider
import java.time.Instant
import kotlin.time.Duration
import kotlin.time.toJavaDuration

/**
 * Controllable clock for tests, previews and UI tests (port of Swift `MockClock`). Thread-safe.
 *
 * `autoAdvance` is added after every [now] read so that successive backend operations get strictly increasing
 * timestamps (like successive Postgres transactions), while staying fully deterministic.
 */
class MockClock(start: Instant, autoAdvance: Duration = Duration.ZERO) {
    private val lock = Any()
    private val step: java.time.Duration = autoAdvance.toJavaDuration()
    private var current: Instant = start

    /** Returns the current date, then advances by `autoAdvance`. */
    fun now(): Instant = synchronized(lock) {
        val value = current
        current = current.plus(step)
        value
    }

    /** The current date, without advancing. */
    fun peek(): Instant = synchronized(lock) { current }

    /** Moves the clock forward (backward with a negative [duration]). */
    fun advance(duration: Duration) {
        val delta = duration.toJavaDuration()
        synchronized(lock) { current = current.plus(delta) }
    }

    fun set(date: Instant) {
        synchronized(lock) { current = date }
    }

    /** A [NowProvider] reading this clock (each call advances by `autoAdvance`). */
    val provider: NowProvider get() = { now() }
}
